### PR TITLE
Improved memory usage in gCNV.

### LINKED ISSUE #5764: Investigate and address memory leak in gCNV.
There appears to be a memory leak in gCNV coming from Theano 0.9.0, possibly fixed in https://github.com/Theano/Theano/pull/5832.  A few possible fixes:

1) Update Theano to the latest 1.0.4 version.  I've tried this and it looks like the leak goes away.  Need to confirm reproducibility of results between versions, see also #5730.
2) Configure Theano 0.9.0 to use MKL, rather than OpenBLAS.  It appears the leak is only an issue with the latter.  This is a little more complicated, since I now realize that MKL is not actually fully utilized (if at all) in our conda environment.  For example, we `pip install numpy`, rather than `conda install` a version from the `default` channel that is compiled against MKL.  So we'd need to change a few dependencies in the environment which might have implications for VQSR-CNN.  See also #4074.  @lucidtronix any thoughts?  @jamesemery and @cmnbroad might also be interested, as this could have pretty drastic implications for the size of the python dependencies---if we go with option 1, we might be able to get rid of MKL, etc.

Not sure if the memory leak manifests the same across all architectures.  Note that I believe this is a separate issue from #5714.

### PR DESCRIPTION
**Pull request by samuelklee**
-Changed sampling of denoised copy ratios to address memory spike and updated output formats and filenames.  Partially addresses #5754.
-Updated theano version to 1.0.4 and changed numpy install source to conda defaults to enable MKL.
-Updated theano flags to use MKL and OpenMP elemwise.

Closes #5764.

### CODE REVIEW

**Comment by samuelklee** [comment]
Prior to this, gCNV shard size and VM type have never been optimized.  For WES cohort mode, we arbitrarily ran with:

37 shards of 200 samples by 5000 intervals on n1-standard-8s (8 CPU, 30GB memory, $0.08 / hr) each taking ~2 hours = ~3 cents / sample

This PR fixes a memory leak and also adds some flags that typically yield a factor of ~2 speedup (notably, the OpenMP elemwise flag), allowing us to run, e.g.: 

2 shards of 50 samples by 100000 intervals on n1-standard-8s (8 CPU, 30GB memory, $0.08 / hr) each taking ~5 hours = ~1.6 cents / sample
4 shards of 50 samples by 50000 intervals on n1-highmem-4s (4 CPU, 26GB memory, $0.05 / hr) each taking ~3.25 hours = ~1.3 cents / sample

The cost was already satisfactory, but fixing the leak allows us to more easily run scatters that are not so wide, which may be crucial for running the megaWDL.  Adding the OpenMP flag also lets CPU scalability work as intended.

Note that we have still not optimized inference.  @mwalker174 @vruano for WGS with 2kb bins, I would expect the cost of the gCNV step to be ~10 cents in cohort mode before inference optimizations, assuming we address #5716 to minimize disk costs.

@asmirnov239 can you review?  And maybe you can address dCR output in PostprocessGermlineCNVCalls and expose the number of samples in a separate PR?  We can make some further changes to the dCR format there if we need.

**Comment by samuelklee** [comment]
@lucidtronix @cmnbroad @jamesemery also note that I went ahead and switched over to numpy MKL, I'm assuming you have no objections!

File: src/main/python/org/broadinstitute/hellbender/gcnvkernel/models/commons.py [hunk 1]
```diff
@@ -226,16 +226,18 @@ def add_sample_to_cum_sum(posterior_sample, _cum_sum):
...
+        model_approx: an instance of PyMC3 mean-field approximation
+        node: a stochastic node in the model
         num_samples: number of samples to draw
 
     Returns:
         A generator that will yield `num_samples` samples from an approximation to a posterior
     """
-    return (model_approx.sample()[model_var_name] for _ in range(num_samples))
+
+    sample = stochastic_node_mean_symbolic(model_approx, node, size=1)
```
Comment (asmirnov239) [inline src/main/python/org/broadinstitute/hellbender/gcnvkernel/models/commons.py]: Can you try using approximation.sample_node method instead?

Comment (samuelklee) [reply src/main/python/org/broadinstitute/hellbender/gcnvkernel/models/commons.py]: Reminder to self to recheck for memory spike as we discussed.

**Comment by samuelklee** [comment]
No spike and the results are exactly the same when using sample_node.  Sorry for not realizing that the other method cribbed from this earlier.

Interestingly, I also checked whether compilation affects memory usage---it doesn't.  Compilation can take a non-trivial amount of time for short-running shards, so we might experiment with distributing a precompiled model for fixed shard sizes in production.  Not sure if this works in practice or what the effect of getting VMs on different architectures might be, though.

mem

**Comment by samuelklee** [comment]
Thanks @asmirnov239, back to you!

**Comment by asmirnov239** [comment]
Looks good @samuelklee! Thanks for fixing it

**Pull request merged**